pkg/providers/tf/cloud-storage.go

```diff
@@ -0,0 +1,191 @@
+
+	"github.com/GoogleCloudPlatform/gcp-service-broker/brokerapi/brokers/models"
+	"github.com/GoogleCloudPlatform/gcp-service-broker/pkg/broker"
+	"github.com/GoogleCloudPlatform/gcp-service-broker/pkg/validation"
+	"github.com/GoogleCloudPlatform/gcp-service-broker/pkg/varcontext"
+	"github.com/pivotal-cf/brokerapi"
+)
+
+func init() {
+	roleWhitelist := []string{
+		"storage.objectCreator",
+		"storage.objectViewer",
+		"storage.objectAdmin",
+	}
+
+	cloudStorage := TfServiceDefinitionV1{
+		Version:          1,
+		Name:             "google-storage-experimental",
+		Id:               "68d094ae-e727-4c14-af07-ee34133c8dfb",
+		Description:      "Experimental Google Cloud Storage that uses the Terraform back-end and grants service accounts IAM permissions directly on the bucket.",
+		DisplayName:      "Experimental Google Cloud Storage",
+		ImageUrl:         "",
+		DocumentationUrl: "",
+		SupportUrl:       "",
+		Tags:             []string{"preview", "gcp", "terraform", "storage"},
+		Plans: []broker.ServicePlan{
+			{
+				ServicePlan: brokerapi.ServicePlan{
+					ID:   "e1d11f65-da66-46ad-977c-6d56513baf43",
+					Name: "standard",
+					Metadata: &brokerapi.ServicePlanMetadata{
+						DisplayName: "Standard",
+					},
+					Description: "Standard storage class.",
+				},
+				ServiceProperties: map[string]string{
+					"storage_class": "STANDARD",
+				},
+			},
+		},
+		ProvisionSettings: TfServiceDefinitionV1Action{
+			PlanInputs: []broker.BrokerVariable{
+				{
+					FieldName: "storage_class",
+					Type:      broker.JsonTypeString,
+					Details:   "The storage class of the bucket. See: https://cloud.google.com/storage/docs/storage-classes.",
+					Required:  true,
+				},
+			},
+			UserInputs: []broker.BrokerVariable{
+				{
+					FieldName: "name",
+					Type:      broker.JsonTypeString,
+					Details:   "The name of the bucket. There is a single global namespace shared by all buckets so it MUST be unique.",
+					Default:   "pcf_sb_${counter.next()}_${time.nano()}",
+					Constraints: validation.NewConstraintBuilder(). // https://cloud.google.com/storage/docs/naming
+											Pattern("^[A-Za-z0-9_\\.]+$").
+											MinLength(3).
+											MaxLength(222).
+											Build(),
+				},
+				{
+					FieldName: "location",
+					Type:      broker.JsonTypeString,
+					Default:   "US",
+					Details:   `The location of the bucket. Object data for objects in the bucket resides in physical storage within this region. See: https://cloud.google.com/storage/docs/bucket-locations`,
+					Constraints: validation.NewConstraintBuilder().
+						Pattern("^[A-Za-z][-a-z0-9A-Z]+$").
+						Examples("US", "EU", "southamerica-east1").
+						Build(),
+				},
+			},
+			Computed: []varcontext.DefaultVariable{
+				{Name: "labels", Default: "${json.marshal(request.default_labels)}", Overwrite: true},
+			},
+			Template: `
+	variable name {type = "string"}
```

Comment:
Having blocks of in-line TF templates like this makes me concerned that as TF becomes the primary source of provisioning for the SB, these in-line templates end up being a growing hole in the test coverage for the project. Consider splitting the templates out and rigging up unit tests for each template which validates the syntax.

Reply:
Discussed this off-line with @josephlewis42 and he clarified that these templates do get utilized/validated and during execution of the examples during integration testing.

Reply:
I have another PR into this branch for this issue #356; it validates that the HCL is syntactically valid (HCL 1.0), and that the inputs and outputs match up with what's expected based on what the plugin author has listed.

Outputs must be an exact match, while the listed inputs can be a superset (some inputs could be deprecated or used in computation, but never by the actual Terraform definition).